src/org/rascalmpl/parser/uptr/recovery/ToTokenRecoverer.java

```diff
@@ -103,35 +107,50 @@ private DoubleArrayList<AbstractStackNode<IConstructor>, AbstractNode> reviveNod
 			for (int j = prods.size() - 1; j >= 0; --j) {
 				IConstructor prod = prods.get(j);
 				
-				IConstructor type = ProductionAdapter.getType(prod);
+				IConstructor sort = ProductionAdapter.getType(prod);
 
-				List<SkippingStackNode<IConstructor>> skippingNodes =
-				findSkippingNodes(input, location, recoveryNode, prod, startLocation);
+				List<SkippingStackNode<IConstructor>> skippingNodes = findSkippingNodes(input, location, recoveryNode, prod, startLocation);
 				for (SkippingStackNode<IConstructor> skippingNode : skippingNodes) {
 					int skipLength = skippingNode.getLength();
-
-					if (!skippedIds.add(Triple.ofNonNull(startLocation, type, skipLength))) {
-						// Do not add this skipped node if a node with the same startLocation, type, and skipLength has been added already
-						continue;
+					Triple<IConstructor, Integer, Integer> key = Triple.of(sort, startLocation, skipLength);
+
+					// See if we already have added a node with this constructor, start location, and skip length
+					SkippingStackNode<IConstructor> previouslyAddedNode = addedNodes.get(key);
+					if (previouslyAddedNode == null) {
+						// No, add the current node
+						addedNodes.put(key, skippingNode);
+						skippingNode.initEdges();
+						recoveredNodes.add(skippingNode, skippingNode.getResult());
+					} else {
+						// We already added a similar node, use that instead
+						skippingNode = previouslyAddedNode;
 					}
 
 					AbstractStackNode<IConstructor> continuer = new RecoveryPointStackNode<>(stackNodeIdDispenser.dispenseId(), prod, recoveryNode);
-
 					EdgesSet<IConstructor> edges = new EdgesSet<>(1);
 					edges.add(continuer);
-				
 					continuer.setIncomingEdges(edges);
 
-					skippingNode.initEdges();
 					skippingNode.addEdges(edges, startLocation);
-					recoveredNodes.add(skippingNode, skippingNode.getResult());
 				}
 			}
 		}
 		
 		return recoveredNodes;
 	}
 
+	// Add a new SkippingStackNode, but only if another node with the same result does not already exist
+	private void addSkippingStackNode(List<SkippingStackNode<IConstructor>> nodes, IConstructor prod, int startLocation, SkippedNode result) {
+		for (SkippingStackNode<IConstructor> node : nodes) {
+			if (((SkippedNode)node.getResult()).getLength() == result.getLength()) {
```

Comment:
Is just comparing the length enough here?

Reply:
We only add nodes with the same production and the same start position. If the length is also the same, it means we would be adding exactly the same (and thus redundant) node.